### ISSUE TITLE
building=house and building=garage should render in slightly different colours

### ISSUE THREAD

**Issue by sladen** [gravitystorm/openstreetmap-carto#1207]
It is useful to be able to tell the houses from the garages and traditionally slightly different shades have been used for building=house and building=garage/building=garages.  This is because it makes residential areas more readable.

Recently (early January 2015) a change appears to have been made to lighten buildings, and now garages and houses are the same colour.  This change is probably quite recent because some tiles have it and some don't.

I can't see anything explicit in the changelog, so this is probably more likely to be an accidental regression.  Likely from commit https://github.com/gravitystorm/openstreetmap-carto/commit/01acfb1b3a8bc1d6254bfdae91ab875ee695684e

**Comment by pnorman** [comment]
> It is useful to be able to tell the houses from the garages and traditionally slightly different shades have been used for building=house and building=garage/building=garages. This is because it makes residential areas more readable.

It was more just something that was carried forwards for historical reasons.

> Recently (early January 2015) a change appears to have been made to lighten buildings, and now garages and houses are the same colour. This change is probably quite recent because some tiles have it and some don't.
> 
> I can't see anything explicit in the changelog, so this is probably more likely to be an accidental regression. Likely from commit 01acfb1

Not doing the old confusing logic for rendering some types of buildings lighter was intentional.

**Comment by Rovastar** [comment]
I never understood building=house being lighter - i could just about understand garages.
Actually depending on the list of motivations for carto osm having building=yes could be considered rendering lighter (although it make little sense now they are all lighter) to encourage people to actually make a more descriptive building tag.

**Comment by mkoniecz** [comment]
After #490 list of lighter buildings was sensible, though I am not sure whatever lighter sheds and tents are still necessary.

@sladen - can you give an example of location where it would result in really significant improvements?

**Comment by sladen** [comment]
@mkoniecz, I noticed while working on an area of Wuerzburg, Germany where the buildings had been drawn, but the housenumbers not yet collected.  In particular this is an area of low-density detached housing, where there's roughly a 1:1 ratio of houses to double garages---which lots of "randomly" placed:

* http://overpass-turbo.eu/s/6Wz ... Wuerzburg, showing {building,use}=gauge{,s}
* http://osm.org/#map=16/49.7753/9.9488 Wuerzburg
* http://overpass-turbo.eu/s/6WC Nottingham Park Estate
* http://www.openstreetmap.org/#map=16/52.9510/-1.1621 Nottingham Park Estate
* http://osm.org/#map=17/52.92126/-1.20610 Beeston, Nottingham

Having the house numbers allows some differientation (when housenumbers have been added, and at those zoom levels), but without we have what appears to be a sea of blobs.

I can also affirm to what @vincentdephily and @Rovastar have noted, there is a feedback loop between demostrating that OSM has richer data, and people using building=* descriptively.  Indeed, seeing the contrast between building=house vs. building={garage{,s},shed} was the motivator for adding garages and tagging them.

**Comment by daganzdaanda** [comment]
+1 for having more differentiation in buildings. There could be 3 tiers: 

 * lower importance: sheds, garage(s), maybe farm_auxiliary, ...? -- these should have the same lightness as buildings now.
 * standard importance: "yes" and catchall (there is a catchall now, isn't it?) -- just a bit darker.
 * higher importance: church, stadium, hospital?, ...? -- another bit darker.

**Comment by kocio-pl** [comment]
Nice idea. Do you think the lower importance buildings should be rendered like building=roof, but with no transparency, or would you like it to be one more rendering type?

**Comment by daganzdaanda** [comment]
Doesn't building=roof render exactly like all other buildings now? There is no transparency anymore.

I believe we can't really make buildings lighter than they are at the moment, so the current version would be the new "light". The new "normal" would need to be a bit stronger. #1208 is related, but the suggestions there are still very close to the current colour. If we want to differentiate, there should probably be a bit more distance. Maybe a stronger outline for the "normal" buildings is enough?

#490 has a good list of buildings that were rendered light: 
* The building types roof, service, shed, shelter, cabin, storage_tank, tank, support, glasshouse, mobile_home, kiosk, silo, canopy and tent are now light. *

**Comment by sladen** [comment]
Regarding lighter-than-light we don't need to compare those solely against whiteness.  Houses and garages normally occur drawn over landuse=residential, so one can optimise for ensuring contrast in the common case.

So this would be optimising constrast between building=house/apartments vs. building=shed/greenhouse/conservatory/garage/garages/roof vs. building=* over landuse=residential.

**Comment by kocio-pl** [comment]
> Doesn't building=roof render exactly like all other buildings now? There is no transparency anymore.

Oh, that's pity... Some large scale roofs should enable user to see the area under them.

**Comment by vincentdephily** [comment]
One of the goals of the restyling was to move away from building transparency altogether. It makes sense in particular with landuse, which have lots of different hues, and with some buildings being only partially over a particular landuse (that's typical when going from residential to commercial, the area often doesn't follow building outlines exactly).

As far as I know there isn't an easy way to, for example, see highway=service through a building but not landuse=*. You have to render a tinted version of the highway over the building, which is fiddly work.

**Comment by mboeringa** [comment]
> Doesn't building=roof render exactly like all other buildings now? There is no transparency anymore.

> Oh, that's pity... Some large scale roofs should enable user to see the area under them.

> As far as I know there isn't an easy way to, for example, see highway=service through a building but not landuse=*. You have to render a tinted version of the highway over the building, which is fiddly work.

No comment... **building=roof**

![rotterdam_railway_station](https://cloud.githubusercontent.com/assets/7635726/5728182/a527afee-9b68-11e4-96fc-2f0b4337c4e3.jpg)

**Comment by sladen** [comment]
@mboeringa, if you to describe the issue in textual form, this would be useful.

**Comment by mboeringa** [comment]
> @mboeringa, if you were to [edit] the comment and describe the issue in textual form, this would be useful.

It is not an issue, it is a possible solution. I have presented this before here somewhere on the issue tracker. It is at least one possible way of rendering **building=roof** without the need of transparency and all the difficulties in interpretation of mixed colors it causes.

I just used a very fine open neutral grey hatch to represent roofs, allowing you to "see through", and render it on top of everything else.

**Comment by sladen** [comment]
I fear the talk of transparency may have got this side-tracked over the last few months.  In-lieu of other suggestions, moving **building=shed/garage** to the colour of amenities/places_of_worship would restore some form of contrast in residential areas, even if it's the other way around.  This would also make that the buildings more likely to have written text in them (houses with house numbers) have good contrast between fill-colour and text-colour.

**Comment by dieterdreist** [comment]
sent from a phone

> Am 15.06.2015 um 04:15 schrieb Paul Sladen <notifications@github.com>:
> 
> In-lieu of other suggestions, moving building=shed/garage to the colour of amenities/places_of_worship would restore some form of contrast in residential areas, even if it's the other way around.


please not, wouldn't make any sense, if we want to do something we should rather lower the visual impact of garages, not raise it. Place of worship buildings are darker because they are generally important, also historically, while garages are typically less important than the buildings they serve.

**Comment by pnorman** [comment]
I didn't realize this issue was still open.

Changing this is technically fairly easy, so if someone wants to propose a change in a PR they're welcome to, but I'm going to close this since I think we're happy with the current state of building rendering.

If someone needs a residential area with a bunch of houses and garages, http://www.openstreetmap.org/#map=18/49.21593/-122.91540 has both mapped in some detail.

**Comment by pnorman** [comment]
To clarify, I do not consider the current situation a bug. As with any issue, views could change with a convincing demonstration of suitable changes.

**Comment by sladen** [comment]
pnorman: Thanks for the clarification, and the reiteration that this is waiting on a CartoCSS/Pull Request of a workable solution to the lack of contrast.

There's probably a conflict here in how to potentially restore the contrast between:

  1. Keeping fewer /different/ colour-levels (consistency through re-using some existing colour to achieve contrast)
  2. Keeping use clarity (consistency through not confusing semantic meaning with some existing colour)

In these cases, @daganzdaanda  was suggesting three levels of building.  A further possible solution would be to adjust the rendering of the *outline* of the buildings; so perhaps rendering lower priority "background" buildings with just the fill and without a contrast on the edge.

The later might possibly meet both of the seemingly conflicting proposals, whilst also re-achieving some level of contrast based on the usage classification.

**Comment by sladen** [comment]
Any potential fix can probably fix issue #738 (`building=greenhouse`) too.

**Comment by vincentdephily** [comment]
I worked toward improving the set and styling of "minor" buildings in the past, and still think that osm-carto should distinguish them. But I don't have time to write a PR at the moment :/

**Comment by danstowell** [comment]
I have found this issue because I'm mapping [a research facility](http://www.openstreetmap.org/#map=18/47.97135/11.23603) with numerous building=shed. The lack of differentiation between substantial buildings and less-substantial structures such as sheds makes the rendering a bit hard to visually scan, in my opinion.

I remembered that garages were rendered lighter, a detail I liked, which is why I came to search here - though IIUC the differentiation is not in the current render at all (not for garages either). So this is just one vote from me with an example, for a mildly-differentiated rendering for auxiliary buildings.

**Comment by d1g** [comment]
@danstowell while I agree that distinction would be nice, I don't think it is possible with a long tali of the current building=* tag or with only one tag:

Peter made some analysis in the past (2012)) of the long tali and possible solutions:
https://wiki.openstreetmap.org/wiki/Talk:Key:building#Current_tag_values
https://wiki.openstreetmap.org/wiki/Talk:Key:building#Adjustments

Note that his definitions are actually belong to http://taginfo.openstreetmap.org/keys/building%3Ause tag, not a building=* tag (a building typology).

Instead, we can render building:levels=1 ...2 buildings differently from others. This approach is easier to add in database and to implement and use/read (not to classify 30-100 building typology tags)
